### PR TITLE
mds: use explicitly sized types for network and disk encoding

### PR DESCRIPTION
**Pull request by lxbsz**
The size of 'unsigned' type maybe not different from different OSes. And we should always use explicitly sized type.

Fixes: https://tracker.ceph.com/issues/63552

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by dparmar18** [comment]
Migrator and Mutation codes too can take this change
```
Migrator.h:    export_base_t(dirfrag_t df, mds_rank_t d, unsigned c, uint64_t g) :
Migrator.h:    unsigned pending_children;
Migrator.h:  unsigned num_locking_exports = 0; // exports in locking state (approx_size == 0)
```

```
Mutation.h:    LockOp(SimpleLock *l, unsigned f=0, mds_rank_t t=MDS_RANK_NONE) :
Mutation.h:    mutable unsigned flags;
Mutation.h:  lock_iterator emplace_lock(SimpleLock *l, unsigned f=0, mds_rank_t t=MDS_RANK_NONE) {
```

**Comment by dparmar18** [comment]
and i think some what server code could also be fixed
https://github.com/ceph/ceph/blob/70d8c5b565c9f2897714844486d456ad78b0d24d/src/mds/Server.h#L496-L507

**Comment by lxbsz** [comment]
> and i think some what server code could also be fixed
> 
> https://github.com/ceph/ceph/blob/70d8c5b565c9f2897714844486d456ad78b0d24d/src/mds/Server.h#L496-L507

This should be fine, because we won't do disk encoding directly with them.

**Comment by lxbsz** [comment]
> Migrator and Mutation codes too can take this change
> 
> ```
> Migrator.h:    export_base_t(dirfrag_t df, mds_rank_t d, unsigned c, uint64_t g) :
> Migrator.h:    unsigned pending_children;
> Migrator.h:  unsigned num_locking_exports = 0; // exports in locking state (approx_size == 0)
> ```
> 
> ```
> Mutation.h:    LockOp(SimpleLock *l, unsigned f=0, mds_rank_t t=MDS_RANK_NONE) :
> Mutation.h:    mutable unsigned flags;
> Mutation.h:  lock_iterator emplace_lock(SimpleLock *l, unsigned f=0, mds_rank_t t=MDS_RANK_NONE) {
> ```

@dparmar18 I didn't see anywhere is encoding this for network or disk, did I miss something here ?

File: src/mds/Capability.h [hunk 1]
```diff
@@ -381,7 +381,7 @@ class Capability : public Counter<Capability> {
   ceph_seq_t mseq = 0;
 
   int suppress = 0;
-  unsigned state = 0;
+  uint32_t state = 0;
```
Comment (mchangir) [inline src/mds/Capability.h:384]: ILP32 and LP64 architectures all implement `unsigned` as `unsigned int` a.k.a. `uint32_t`
could you point where this standard isn't being followed ?

however, I agree its always a good idea to specify the data size upfront during serialization/deserialization

Comment (lxbsz) [reply src/mds/Capability.h:384]: I couldn't tell exactly where. I just found this when I was coding the kclient patches and googled about this. It's said in some platform or OS it maybe different. As @gregsfortytwo mentioned we should always use explicitly sized type in these cases.

Comment (mchangir) [reply src/mds/Capability.h:384]: so is `uint32_t` good enough or is `__le32` more appropriate ?

Comment (lxbsz) [reply src/mds/Capability.h:384]: The `uint32_t` is good enough here and the `encode()` will help switch it to `__le32`.

**Comment by dparmar18** [comment]
> > Migrator and Mutation codes too can take this change
> > ```
> > Migrator.h:    export_base_t(dirfrag_t df, mds_rank_t d, unsigned c, uint64_t g) :
> > Migrator.h:    unsigned pending_children;
> > Migrator.h:  unsigned num_locking_exports = 0; // exports in locking state (approx_size == 0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > Mutation.h:    LockOp(SimpleLock *l, unsigned f=0, mds_rank_t t=MDS_RANK_NONE) :
> > Mutation.h:    mutable unsigned flags;
> > Mutation.h:  lock_iterator emplace_lock(SimpleLock *l, unsigned f=0, mds_rank_t t=MDS_RANK_NONE) {
> > ```
> 
> @dparmar18 I didn't see anywhere is encoding this for network or disk, did I miss something here ?

scrolling through these methods i saw usage of fixed size types in their params hence the comment, but since they play no role in encoding, it doesn't make sense to touch them

**Comment by dparmar18** [comment]
jenkins test make check arm64

**Review by dparmar18 — APPROVED** [review]

**Review by batrick — APPROVED** [review]

**Review by rishabh-d-dave — APPROVED** [review]

**Comment by rishabh-d-dave** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#16-Jan-2023

**Pull request merged**